compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3553,6 +3555,70 @@ TR::Register *OMR::Power::TreeEvaluator::vmulInt32Helper(TR::Node *node, TR::Cod
    return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::vmuluwm);
    }
 
+TR::Register *OMR::Power::TreeEvaluator::vmulInt64Helper(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   if (cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P10))
+      return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::vmulld);
+
+   /*
+   To perform the multiplication without the vmulld instruction, we will break it into parts.
+
+   Suppose we have two long integers (64 bits), AB and CD, where each of A, B, C, and D is one word (32 bits).
+   To get their product, AB * CD, we can do the following:
+
+   AB * CD = B*D + 2^32*(AD + BC) + 2^64*(A*C)
+
+   Since we are only keeping the lower 64 bits of the result, this is equivalent to:
+
+   AB * CD = B*D + 2^32*(AD + BC)
+   */
+
+   TR::Node *firstChild;
+   TR::Node *secondChild;
+   TR::Register *lhsReg, *rhsReg;
+   TR::Register *productReg;
+   TR::Register *tempA, *tempB;
+   TR::Register *shiftReg;
+
+   firstChild = node->getFirstChild();
+   secondChild = node->getSecondChild();
+
+   lhsReg = cg->evaluate(firstChild);
+   rhsReg = cg->evaluate(secondChild);
+
+   productReg = cg->allocateRegister(TR_VRF);
+   tempA = cg->allocateRegister(TR_VRF);
+   tempB = cg->allocateRegister(TR_VRF);
+   shiftReg = cg->allocateRegister(TR_VRF);
+
+   node->setRegister(productReg);
+
+   //set shift amount to 32 bits
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::vspltisw, node, shiftReg, -16);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vadduwm, node, shiftReg, shiftReg, shiftReg);
+
+   //productReg = B*D
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vmulouw, node, productReg, lhsReg, rhsReg);
+
+   //tempB = 2^32*(AD + BC)
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vrld, node, rhsReg, rhsReg, shiftReg);
```

Comment:
You can't overwrite `rhsReg` here, since that register might be shared with other evaluators. It looks like `tempB` could be reused for this purpose, since this shifted value is never used again once that register is overwritten.